user docs: Replace right simplebar with browser scrollbar

Signior-X opened the pull request:
This commit removes the simplebar implementation for the right.
Thus using the default browser scrollbar. Removes simplebar class
and functions scrollToHash with this commit.

Fixes #10732 

<!-- What's this PR for?  (Just a link to an issue is fine.) -->


**Testing plan:** <!-- How have you tested? -->
Tested manually and with node-tests.

**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
How it looks now
![image](https://user-images.githubusercontent.com/56730716/107115928-fed94600-6895-11eb-8b12-d6d2051fb7ad.png)


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

andersk commented:
This is an improvement and I’d be happy to merge as-is. But in the issue I [suggested](https://github.com/zulip/zulip/issues/10732#issuecomment-478757652) a further improvement that also addresses other accessibility problems:

> If you make the right scrollbar be the normal scrollbar on `<html>` (rather than the overflow scrollbar of a separate pane), and position the header and left sidebar using `position: fixed`, you won’t have any of those issues.

Are you interested in trying to implement that?

Signior-X commented:
Sure, I will try that!
Let me see, I think for desktop it will work fine for sure. Just need to see for smaller devices.

Signior-X commented:
@andersk done the improvement suggested by you. Here is the gif now, it uses html markdown and also takes care nothing is broken for smaller screen devices. Tested it properly.

![scrollbar](https://user-images.githubusercontent.com/56730716/107126977-138afd80-68d9-11eb-976f-49cfd5cdb07f.gif)

andersk commented on static/styles/portico/portico.css:
```diff
@@ -266,6 +265,11 @@ html {
     background-color: hsl(0, 0%, 100%);
     width: calc(100vw - 300px);
     height: calc(100vh - 59px);
+    overflow: unset;
```
Instead of this override, we can just remove `overflow: auto` from `.markdown` static/styles/portico/markdown.css. The only other use of `.markdown` is http://localhost:9991/devtools/ which doesn’t need the `overflow: auto`.

Signior-X replied:
Sure, that was one of my major concerns. That's why I didn't change there. I will remove the overflow: auto from .markdown

Signior-X replied:
Solved.

Signior-X commented on static/styles/portico/portico.css:
```diff
@@ -266,6 +265,11 @@ html {
     height: calc(100vh - 59px);
+    overflow: unset;
+
+    @media (min-width: 800px) {
```
Oh got it, I will control this from @media max-width block then.

Signior-X replied:
Solved.

Signior-X commented on static/styles/portico/portico.css:
```diff
@@ -1353,9 +1357,14 @@ input.new-organization-button {
 }
 
 @media (width <= 800px) {
+    .app {
+        overflow: hidden;
+    }
+
     .app.help {
         .markdown {
             width: 100vw;
+            overflow: auto;
```
Actually, first I tried without this, but for better view in smaller devices this is needed. 
That is using html scrollbar in small screens too, but that actually disrupts the view as you can see in this picture
![image](https://user-images.githubusercontent.com/56730716/107129741-99b13f00-68ed-11eb-90f8-0c8166ef4e37.png)

Having the overflow auto for that element itself works better for smaller devices and touch-based also. So as you can see in this picture, even becoming small it will not disrupt the view.
![image](https://user-images.githubusercontent.com/56730716/107129900-c9147b80-68ee-11eb-9197-b291dc272b06.png)

And yes simplebar is completely removed from the right scrollbar.

Signior-X replied:
Sure to hear from you what feels better. We can do that way.

andersk replied:
I think the only problem in your “disrupted” screenshot is that the left sidebar is too narrow, right? This seems to be a problem with `100vw` vs. `100%`, and this fixes it:

```diff
--- a/static/styles/portico/portico.css
+++ b/static/styles/portico/portico.css
@@ -1353,15 +1353,10 @@ input.new-organization-button {
 }
 
 @media (width <= 800px) {
-    .app {
-        overflow: hidden;
-    }
-
     .app.help {
         .markdown {
             width: 100vw;
             margin-left: 0;
-            overflow: auto;
 
             .content {
                 max-width: none;
@@ -1383,7 +1378,7 @@ input.new-organization-button {
         .sidebar {
             position: fixed;
             top: 59px;
-            right: calc(100vw - 50px);
+            right: calc(100% - 50px);
             width: 100%;
             min-width: unset;
             height: calc(100vh - 59px);
```

Signior-X replied:
Cool! This works awesome. Got it, making this change quickly.

Signior-X commented:
@andersk done the changes suggested by you.

Signior-X commented:
@andersk now I think all changes done as suggested. Tell what to do.... if this is now fine, should I squash all the commits? or leave it as it is.

timabbott commented:
We should be sure to test that links to headings partway down the page still work with this change?  We have a few of those, easy to find using `git grep '[a-z]#[a-z]' templates/zerver/help/` if you have trouble finding an example.

Signior-X commented:
Yes, I will check that.

Signior-X commented:
I checked it, as I  tried to open with link to headings pathway or change the hash of a page, the id is just some way behind the topbar as shown below
![image](https://user-images.githubusercontent.com/56730716/108239015-7b5e0580-716f-11eb-9c67-955059416823.png)

I think, adding some offset will really help in this. I will try to do this.

Signior-X commented:
Updated it now, it positions it to the right location now!
![image](https://user-images.githubusercontent.com/56730716/108244986-cd098e80-7175-11eb-8908-88e3c3d88af1.png)

andersk commented:
Don’t use JavaScript to force in this offset. This is what the [`scroll-margin-top`](https://developer.mozilla.org/en-US/docs/Web/CSS/scroll-margin-top) property is for.

Signior-X commented:
Thanks @andersk updated :+1:

Signior-X commented:
> Don’t use JavaScript to force in this offset. This is what the [`scroll-margin-top`](https://developer.mozilla.org/en-US/docs/Web/CSS/scroll-margin-top) property is for.

@andersk I have removed the js and added the `scroll-margin-top`.
Can you review this now?

timabbott commented:
This PR seems to introduce some weird artifacts (at least I saw that on localhost and not on zulip.com):

![image](https://user-images.githubusercontent.com/2746074/111528661-aa3fab00-871e-11eb-9635-8a54a8d0e7c4.png)

Signior-X commented:
I looked over this. The issue is caused by this line in `portico.css`
```css
body {
    background-color: hsl(0, 0%, 98%);
}
```
Adding a screenshot with this property and without this.
![image](https://user-images.githubusercontent.com/56730716/111549065-99844900-87a1-11eb-946f-a4ea3829c4ae.png)

![image](https://user-images.githubusercontent.com/56730716/111549186-c6d0f700-87a1-11eb-933b-ff0d7e662a50.png)

Removing this fixes this grey area. However, I am not user that should it be removed or not.

Signior-X commented:
@timabbott fixed the grey square space. Check this one now.

timabbott commented:
That fixed that issue, but it created this new issue:

![image](https://user-images.githubusercontent.com/2746074/112682480-64c16300-8e2d-11eb-8c23-f0350323a462.png)

Probably depends on window height, but it appears that the grey area below the window is too tall and results in a scrollbar on pages that should not have one.

Also, another bug is that if you're scrolled down and click on another /help/ page, it doesn't take you to the top of the page you clicked to.

Signior-X commented:
This PR is getting hard to track also due to changing a css, getting new issue created.
I think first, I should split this PR into two commits.
First one -> Removing the simplbar and usign scrollbar
second one -> Then trying to change it to html as @andersk suggested
> If you make the right scrollbar be the normal scrollbar on <html> (rather than the overflow scrollbar of a separate pane), and position the header and left sidebar using position: fixed, you won’t have any of those issues.

This will help in easily tracking issues. I am doing this now.

Signior-X commented:
Now the first commit is working properly. Tested it for patches, links going, hash working, and scrolling to top. This actually fixes the issue by replacing the simplebar with the browser scrollbar which scrolls according to `.markdown` element.

Now as some changes suggested by @andersk , we should try to remove the scroll from `.markdown` to the `html`. Just a better version of this fix. That is getting a bit tricky and making css fixes resulting in some or other form of distortion.
So, now trying to do this over the first commit. Hope to complete this by today.

Signior-X commented on static/js/portico/help.js:
```diff
@@ -65,14 +65,10 @@ function render_code_sections() {
-        scrollbar.scrollTop = position;
-    } else {
-        scrollbar.scrollTop = 0;
+function scroll_top_if_empty_hash(ele) {
```
If the page content is updated and the hash remains empty, this function scrolls the page to the top. If there is any hash in the link, it is automatically handled by the browser scrollbar. So, no need to add that case.

Signior-X commented on static/styles/portico/markdown.css:
```diff
@@ -62,6 +61,7 @@
     h3 {
         font-weight: 700;
         user-select: none;
+        scroll-margin-top: 70px;
```
This adds a scroll margin for the element so that when the links are opened using hash, it correctly positions the page.

Signior-X commented on static/styles/portico/portico.css:
```diff
@@ -2,7 +2,6 @@ body {
     background-color: hsl(0, 0%, 98%);
     font-family: "Source Sans 3", Helvetica, Arial, sans-serif;
     line-height: 150%;
-    height: 100%;
```
The body height should be removed, as this isn't fixed now and is automatic handles by the content on the page. Having this line, the page where the scrollbar was not needed was also having one scrollbar.
Fixes this bug.

Signior-X commented on static/styles/portico/portico.css:
```diff
@@ -151,7 +149,7 @@ html {
 
         border-right: 1px solid hsl(219, 10%, 97%);
-        overflow: auto;
+        position: fixed;
```
Make the sidebar fixes, so that the html scrollbar does not scrolls this and sidebar remains fixed. it has its own simplebar.

Signior-X commented on static/styles/portico/portico.css:
```diff
@@ -265,7 +263,7 @@ html {
     background-color: hsl(0, 0%, 100%);
     width: calc(100vw - 300px);
-    height: calc(100vh - 59px);
+    margin-left: 300px;
```
The default height is removed from the .markdown as now the scroll is handled by the html, not the .markdown. This solves the issue of some wried grey patches coming.

Signior-X commented on static/styles/portico/portico.css:
```diff
@@ -1377,7 +1376,7 @@ input.new-organization-button {
             position: fixed;
             top: 59px;
-            right: calc(100vw - 50px);
+            right: calc(100% - 50px);
```
Explained in the commits as well as in the conversation as the browser scroll bar width should not be considered while calculating the right position.

Signior-X commented on static/styles/portico/portico.css:
```diff
@@ -1356,6 +1354,7 @@ input.new-organization-button {
     .app.help {
         .markdown {
             width: 100vw;
+            margin-left: 0;
```
Making mobile friendly.

Signior-X commented:
@timabbott I have properly tested and rebased the PR. Also fixed the problems as mentioned by you in the comments.
Also, added a single comment in the PR, so that it becomes easy to understand why that css was used.

Also tested it properly or links opening, and height of elements. The default height of every element like body, markdown is now removed. And the function to handle the scroll for the update page is created. I have tried to keep the commits as coherent as possible now. It would be really nice if this can checked now.

Signior-X commented on static/styles/portico/portico.css:
```diff
@@ -1397,7 +1396,7 @@ input.new-organization-button {
 
             + .hamburger {
-                transform: translateX(calc(100vw - 50px));
+                transform: translateX(calc(100vw - 72px));
```
@vinitS101 as pointed by you. This was happening.
![image](https://user-images.githubusercontent.com/56730716/119663233-cdd64e80-be4f-11eb-9cf4-daaf834646c1.png)

This shifts a bit to left so that there will be no overlap in anycase.
![image](https://user-images.githubusercontent.com/56730716/119663328-e21a4b80-be4f-11eb-9d3b-7e72982c0d05.png)
Also on mobile devices, it looks fine and fixes both the problems.

Signior-X commented:
@vinitS101 this one is now ready for review.

Signior-X commented:
The last commit https://github.com/zulip/zulip/pull/17227/commits/d1c84452fc722bcffb62bfabec52810841062d9c fixes  the problem that we discussed of the alignment of hamburger icon and Home. This fixes this one.
This PR is now ready for another review.

vinitS101 commented:
Tested this locally, changes to the scrollbar look good to me.

Signior-X commented:
Just rebased and updated.

andersk reviewed (approved):
Rebased and simplified, largely due to #15713.

timabbott commented:
Merged, thanks @Signior-X and @andersk!
